### PR TITLE
HSEARCH-3597 Document all available predicates/sorts/projections and query DSL basics

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-3597

Based on #2012, which should be merged first.

The easiest way to review this PR is probably to build the documentation and then read the "Searching" section:

```
mvn clean install -DskipTests -pl documentation -am
xdg-open ./documentation/target/asciidoctor/en-US/html_single/index.html
```

### CODE REVIEW

**Comment by yrodiere** [comment]
Rebased on master just now.

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 1]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
-include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/directfieldmapping/HibernateOrmSimpleMappingIT.java[tags=projection-simple-objects]
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=bool-and]
 ----
+<1> The hits *must* have a `title` field matching the text `robot`,
+independently from other clauses in the same boolean predicate.
+<2> The hits *must* have a `description` field matching the text `crime`,
+independently from other clauses in the same boolean predicate.
+<3> All returned hits will match *all* of the clauses above:
+they will have `title` field matching the text `robot`
+*and* they will have a `description` field matching the text `crime`.
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: Same here with `crime`.

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 2]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=bool-mustNot]
+----
+<1> The hits *must* have a `title` field matching the text `robot`,
+independently from other clauses in the same boolean predicate.
+<2> The hits *must not* have a `description` field matching the text `murder`,
+independently from other clauses in the same boolean predicate.
+<3> All returned hits will match *all* of the clauses above:
+they will have `title` field matching the text `robot`
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]:
```suggestion
they will have a `title` field matching the text `robot`
```

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 3]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
 ----
-include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/directfieldmapping/HibernateOrmSimpleMappingIT.java[tags=projection-simple-objects]
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=bool-and]
 ----
+<1> The hits *must* have a `title` field matching the text `robot`,
+independently from other clauses in the same boolean predicate.
+<2> The hits *must* have a `description` field matching the text `crime`,
+independently from other clauses in the same boolean predicate.
+<3> All returned hits will match *all* of the clauses above:
+they will have `title` field matching the text `robot`
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]:
```suggestion
they will have a `title` field matching the text `robot`
```

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 4]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
+<1> The hits *must* have a `title` field matching the text `robot`,
+independently from other clauses in the same boolean predicate.
+<2> The hits *should* have a `description` field matching the text `crime`,
+but they may not, because matching the `must` clause above is enough.
+However, matching this `should` clause will improve the score of the document.
+<3> The hits *should* have a `description` field matching the text `murder`,
+but they may not, because matching the `must` clause above is enough.
+However, matching this `should` clause will improve the score of the document.
+<4> All returned hits will match the `must` clause, and optionally the `should` clauses:
+they will have `title` field matching the text `robot`,
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]:
```suggestion
they will have a `title` field matching the text `robot`,
```

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 5]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
 ====
 
-The projection DSL offers more projection types, and multiple options for each type of projection.
-To learn more about the field projection, and all the other types of projection,
-refer to the following sections.
+.Matching a simple query string: NOT operator
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=simpleQueryString-not]
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: Do you need the `+` here?

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 6]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
+[[search-dsl-predicate-spatial-within]]
+=== `within`: match points within a circle, box, polygon
+// Search 5 anchors backward compatibility
+[[spatial-queries]]
+
+.Matching points within a circle
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=within-circle]
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: Maybe something more realistic than a circle of 5000 km would be better?

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 7]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
+----
+====
+
+// TODO alternative syntaxes?
+
+.Matching points within a polygon
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/PredicateDslIT.java[tags=within-box]
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]: You don't point to the right example.

File: documentation/src/main/asciidoc/search-dsl.asciidoc [hunk 8]
```diff
@@ -85,149 +120,1023 @@ include::todo-placeholder.asciidoc[]
...
+or a `org.hibernate.search.util.common.function.TriFunction`.
+It will combine values returned by other projections and create an object returned by the composite projection.
+Depending on the type of function,
+either one, two, or three additional arguments are expected.
+<3> Define the first projection to combine as a projection on the `title` field,
+meaning the constructor of `MyPair` will be called for each matched document
+with the value of the `title` field as its first argument.
+<4> Define the second projection to combine as a projection on the `genre` field,
+meaning the constructor of `MyPair` will be called for each matched document
+with the value of the `genre` field as its first argument.
```
Comment (gsmet) [inline documentation/src/main/asciidoc/search-dsl.asciidoc]:
```suggestion
with the value of the `genre` field as its second argument.
```

File: documentation/src/main/asciidoc/search-dsl.asciidoc (hunk 5 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/search-dsl.asciidoc]: If I don't use `withAndAsDefaultOperator`, yes. And I'd rather have one example for each feature.

**Comment by yrodiere** [comment]
Merging. Thanks!

**Pull request merged**